checkmedia.c

```diff
@@ -1,598 +1,205 @@
-#define _GNU_SOURCE
-
 #include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-#include <unistd.h>
-#include <ctype.h>
-#include <fcntl.h>
-#include <stdint.h>
-#include <sys/types.h>
-#include <sys/stat.h>
-
-#include "md5.h"
-#include "sha1.h"
-#include "sha256.h"
-#include "sha512.h"
-
-#define MAX_DIGEST_SIZE SHA512_DIGEST_SIZE
-
-typedef enum {
-  digest_none, digest_md5, digest_sha1, digest_sha224, digest_sha256, digest_sha384, digest_sha512
-} digest_t;
-
-typedef union {
-  struct md5_ctx md5;
-  struct sha1_ctx sha1;
-  struct sha256_ctx sha224;
-  struct sha256_ctx sha256;
-  struct sha512_ctx sha384;
-  struct sha512_ctx sha512;
-} digest_ctx_t;
-
-
-static void do_digest(char *file);
-static void get_info(char *file, unsigned opt_verbose);
-static char *no_extra_spaces(char *str);
-static void update_progress(unsigned size);
-static uint32_t read_le32(const void *ptr);
-static void check_mbr(const unsigned char *mbr);
-
-static void digest_media_init(digest_ctx_t *ctx);
-static void digest_media_process(digest_ctx_t *ctx, unsigned char *buffer, unsigned len);
-static void digest_media_finish(digest_ctx_t *ctx, unsigned char *buffer);
+#include <getopt.h>
+
+#include "mediacheck.h"
 
+void help(void);
+int progress(unsigned percent);
 
 struct {
-  struct {
-    digest_ctx_t ctx;
-    digest_ctx_t full_ctx;
-    digest_t type;                              /* digest type */
-    char *name;                                 /* digest name */
-    int size;                                   /* digest size */
-    unsigned got_old:1;                         /* got digest stored in iso */
-    unsigned got_current:1;                     /* calculated current digest */
-    unsigned ok:1;                              /* digest matches */
-    unsigned char old[MAX_DIGEST_SIZE];         /* digest stored in iso */
-    unsigned char current[MAX_DIGEST_SIZE];     /* digest of iso ex special area */
-    unsigned char full[MAX_DIGEST_SIZE];        /* full digest of iso */
-  } digest;
-
-  unsigned err:1;		/* some error */
-  unsigned err_ofs;		/* read error pos */
-  unsigned size;		/* in kb */
-  unsigned hybrid_size;		/* hybrid partition, in kb */
-  unsigned media_nr;		/* media number */
-
-  char *media_type;		/* media type */
-  char vol_id[33];		/* volume id */
-  char app_id[81];		/* application id */
-  char app_data[0x201];		/* app specific data*/
-  unsigned pad;			/* pad size in kb */
-  unsigned pad_is_skip:1;	/* skip last sectors */
-} iso;
+  unsigned verbose:1;
+  unsigned help:1;
+  unsigned version:1;
+  char *file_name;
+} opt;
+
+struct option options[] = {
+  { "help", 0, NULL, 'h' },
+  { "verbose", 0, NULL, 'v' },
+  { "version", 0, NULL, 1 },
```

Comment:
I know there is a collision of `'v'` but this looks a bit inconsistent to me. Can be C define used here so it makes code easier to read?

like

```
#DEFINE HELP_OPTION 'h'
#DEFINE VERBOSE_OPTION 'v'
#DEFINE VERSION_OPTION 'o'

struct option options[] = {
  { "help", 0, NULL, HELP_OPTION },
  { "verbose", 0, NULL, VERBOSE_OPTION },
  { "version", 0, NULL, VERSION_OPTION },
  { }
};
...
// and later in switch
switch(i) {
     case VERSION_OPTION:
...

Reply:
Come on, it's just 3 options in total and all the getopt parsing is within a single terminal screen.